lib/classes/Variables.js

```diff
@@ -167,23 +168,33 @@ class Variables {
   }
 
   getValueFromSource(variableString) {
+    const cached = this.cache[variableString];
+    if (cached) {
+      return cached;
+    }
+    let valueFromSource;
     if (variableString.match(this.envRefSyntax)) {
-      return this.getValueFromEnv(variableString);
+      valueFromSource = this.getValueFromEnv(variableString);
     } else if (variableString.match(this.optRefSyntax)) {
-      return this.getValueFromOptions(variableString);
+      valueFromSource = this.getValueFromOptions(variableString);
     } else if (variableString.match(this.selfRefSyntax)) {
-      return this.getValueFromSelf(variableString);
+      valueFromSource = this.getValueFromSelf(variableString);
     } else if (variableString.match(this.fileRefSyntax)) {
-      return this.getValueFromFile(variableString);
+      valueFromSource = this.getValueFromFile(variableString);
     } else if (variableString.match(this.cfRefSyntax)) {
-      return this.getValueFromCf(variableString);
+      valueFromSource = this.getValueFromCf(variableString);
     } else if (variableString.match(this.s3RefSyntax)) {
-      return this.getValueFromS3(variableString);
+      valueFromSource = this.getValueFromS3(variableString);
     } else if (variableString.match(this.stringRefSynax)) {
-      return this.getValueFromString(variableString);
+      valueFromSource = this.getValueFromString(variableString);
     } else if (variableString.match(this.ssmRefSyntax)) {
-      return this.getValueFromSsm(variableString);
+      valueFromSource = this.getValueFromSsm(variableString);
+    }
+    if (valueFromSource) {
+      this.cache[variableString] = valueFromSource;
+      return valueFromSource;
```

Comment:
Awesome. I have just updated this - so that once the promise is resolved it replaces itself in the cache with the resolved value.

I had similar thoughts to you when initial writing this - that ideally you should not cache a promise. But by caching the promise, the promise itself takes care of the `waiters` array in your pseudo code. As while the promise is yet to be resolved, any new call for the variable gets the promise that will eventually become the resolved value. Also if there are multiple requests, the appended .then()’s are executed in order of attachment - so it the promise acts as its own queue. This ends up doing the same thing as your proposed implementation but with less complexity.